File: 1-analyzing-algorithms/Big-O-notation/4-quadratic_time.js
function checkDuplicates(arr) {
  let steps = 0;
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      steps += 1;
      console.log("step:", steps);
      if (i != j) {
        if (arr[i] == arr[j]) {
          console.log("Found Duplicates", arr[i], arr[j]);
        }
      }
    }
  }
}

let example = [1, 2, 3, 4, 4, 3, 2, 1, 6, 5, 4, 2, 3, 1]; // array size is 14. The function takes 14*14 = 196 steps

checkDuplicates(example);

/*
    Quadratic Time Complexity represents an algorithm whose performance is directly proportional to the squared size of the input data set (similar to linear, but squared). 
    Within programs, this time complexity occurs whenever we nest over multiple iterations within the data sets.

    The example above is an example for O(n^2) time complexity (although probably the function itself does not make sense :)). 
    For each element in the array, the function starts another loop that goes through every other element in the array and thus create the quadratic runtime. 

*/

File: 1-Time_Complexity/4-quadratic_time.py
def quadratic_time(arr):
    counter = 0
    for val in arr:
        for val in arr:
            counter += 1
    print(counter)


arr = [1, 2, 3, 4, 5]

quadratic_time(arr)

'''
Quadratic Time Complexity refers to the cases in which the execution time of an algorithm is proportional to the squared size of the input data set (similar to linear, but squared). 
This time complexity usually occurs when there are multiple for-loops nested. 

The example above is an example for O(n^2) time complexity (although probably the function itself does not make sense :)). 
The function iterates through each element of the array. In each iteration, another for loop gets executed that goes through every other item in the array. 
This double iteration leads to the quadratic runtime.  
'''
